### ISSUE TITLE
-std=c99 / gnu99

### ISSUE THREAD

**Issue by rickygraham** [pure-data/pd-lib-builder#15]
I ran into a snag when using lib-builder to compile an external:

```
grambidec.c:367:5: error: ‘for’ loop initial declarations are only allowed
in C99 mode
    for(int i = 0; i < 7; i++) {
    ^
grambidec.c:367:5: note: use option -std=c99 or -std=gnu99 to compile your
code
```

Do we need to include c99 / gnu99 CFLAG in the Makefile?

**Comment by residuum** [comment]
Yes, you must add all your cflags to the `cflags` variable, e. g. from my own external:

    cflags += -DPD -DVERSION='"$(lib.version)"' -I"$(uthash)" -std=c99 
    
    (lots of other stuff)
    
    include pd-lib-builder/Makefile.pdlibbuilder

**Comment by umlaeute** [comment]
hmm, yes and no.

the `Makefile.pdlibbuilder` will include any *standard* flags, needed to compile an external.
for instance, you do not need to (nor should you) add flags for the search path to Pd's headers, nor do you need to add `-DPD` (though in about 95% of all externals I know this define is never actually used anyhow).

however, if you have any *special needs* for you external (e.g. using a 3rd party library that needs special includes and/or libraries to link against; or using a C-standard that is not the default for most current compilers), then you can (and should) add these flags to manually.

so @residuum's example should be more like:

     cflags += -DVERSION='"$(lib.version)"' -I"$(uthash)" -std=c99

and you should only try to add a minimum set of flags (and only flags that enable a minimum and most standardized set of required features), e.g. use `-std=c99` instead of ~~`-std=gnu99`~~ for portability sake.

**Comment by umlaeute** [comment]
@rickygraham does this resolve your problem (and can the issue be closed)?

**Comment by katjav** [comment]
Today I came across info that shines a different light on this issue (for me at least). Earlier I thought it was related to really outdated gcc versions, but no...

Gcc has used -std=gnu89 as default for a long time, even when -std=c99 and -std=gnu99 were supported. But, the default jumped from -std=gnu89 to -std=gnu11 with gcc 5. That was only last year!

https://gcc.gnu.org/gcc-5/changes.html

This is going to cause more than incidental confusion I'm afraid. @rickygraham encountered the issue after he developed in Xcode, and used pd-lib-builder to build with Linux and MinGW. It will be even more confusing if someone else than the lib developer is the first to be confronted with such errors.

What can we do? At least document the situation on a very visible place.
